src/V3Timing.cpp

```diff
@@ -85,18 +90,144 @@ class TimingVisitor final : public VNVisitor {
     };
 
     // NODE STATE
-    //  AstNode::user1()                         -> bool.         Set true if the node has been
+    //  AstClass::user1()                        -> bool.         Set true if the class member
+    //                                                            cache has been refreshed.
+    //  Ast{NodeProcedure,CFunc,Begin}::user2()  -> bool.         Set true if process/task is
+    //                                                            suspendable
+    //  Ast{NodeProcedure,CFunc,Begin}::user3()  -> DependencyVertex*.  Vertex in m_depGraph
+    const VNUser1InUse m_user1InUse;
+    const VNUser2InUse m_user2InUse;
+    const VNUser3InUse m_user3InUse;
+
+    // STATE
+    AstClass* m_classp = nullptr;  // Current class
+    AstNode* m_procp = nullptr;  // NodeProcedure/CFunc/Begin we're under
+    V3Graph m_depGraph;  // Dependency graph where a node is a dependency of another if it being
+                         // suspendable makes the other node suspendable
+
+    // METHODS
+    // Get or create the dependency vertex for the given node
+    DependencyVertex* getDependencyVertex(AstNode* const nodep) {
+        if (!nodep->user3p()) nodep->user3p(new DependencyVertex{&m_depGraph, nodep});
+        return nodep->user3u().to<DependencyVertex*>();
+    }
+    // Create a dependency edge between the given nodes
+    void makeDependencyEdge(DependencyVertex* const fromVxp, DependencyVertex* const toVxp) {
+        for (V3GraphEdge* edgep = fromVxp->outBeginp(); edgep; edgep = edgep->inNextp()) {
+            if (edgep->top() == toVxp) return;
+        }
```

Comment:
This seems O(n^2) on edges.  Perhaps better to just make the edge and then use the graph algorithm to delete the duplicates?

Reply:
We also propagate during AST traversal, so if we don't limit deduplicate edges here, we lose time on propagation. Perhaps it would be faster to build the graph, delete duplicates, and then propagate.